### PR TITLE
Feature/login

### PR DESCRIPTION
**Pull request by andriy-kozlytski**
https://app.asana.com/0/1200443148768940/1201699484353283/f

### CODE REVIEW

File: public/js/authorization.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    var jsonPayload = decodeURIComponent(atob(base64).split('').map(function(c) {
+        return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
+    }).join(''));
+
+    return JSON.parse(jsonPayload);
+};
+
+export function logout () {
+	// request('/auth/logout');
+	fetch('/auth/logout');
```
Comment (lodry) [inline public/js/authorization.js]: перевіряй результат

File: server.php [hunk 2]
```diff
@@ -18,18 +20,21 @@
...
 {
-    $mimeType = $fileinfo->file($path, FILEINFO_MIME_TYPE);
+    // $mimeType = $fileinfo->file($path, FILEINFO_MIME_TYPE);
+    $mimeType = getFileMimeType($path);
     $res->setHeader("Content-Type", $mimeType);
-    $file = file_get_contents_async($path, fn($arg)=>var_dump($arg)&$res->setHeader("Content-Length", strlen($arg))
+    var_dump($mimeType);
+    // $file = file_get_contents_async($path, fn($arg)=>var_dump($arg)&$res->setHeader("Content-Length", strlen($arg))
+        // ->send($arg)); // FIX PROBLEM WITH ASYNC READ
+    file_get_contents_async($path, fn($arg)=>$res->setHeader("Content-Length", strlen($arg))
```
Comment (lodry) [inline server.php]: типізуй функцію

File: src/Controllers/AuthController.php [hunk 3]
```diff
@@ -15,23 +17,167 @@ public function __construct(
...
+                (object)['error' => "User already exist"]
+            ));
+            return;
+        }
+        $userData = $this->userService->register($candidate);
+        var_dump('userdata', $userData);
+        $this->logger->info('Register user', ['email' => $candidate->getEmail()]);
+        // TODO resolve problem Max age
+        $maxAge = 3600 * 24 * 15;
+        $response->setHeader("Set-Cookie",'refreshToken=' . $userData->tokens->refreshToken . '; Path=/; Max-Age=' . $maxAge . '; HttpOnly');
```
Comment (lodry) [inline src/Controllers/AuthController.php:56]: повторно воно не ставить

Comment (andriy-kozlytski) [reply src/Controllers/AuthController.php:56]: Did you check it?
I check it works
and it is alike native behaviour

File: server.php (hunk 2 above)
Comment (andriy-kozlytski) [reply server.php]: I'm not sure
check me

File: src/Controllers/AbstractController.php [hunk 4]
```diff
@@ -4,11 +4,23 @@
...
-    private readonly ContainerBuilder $containerBuilder;
+    protected readonly ContainerBuilder $containerBuilder;
 
     public function setContainer(ContainerBuilder $container): self
     {
         $this->containerBuilder = $container;
         return $this;
     }
+
+    public function getCookie(string $rawCookie, string $name): string|bool
```
Comment (lodry) [inline src/Controllers/AbstractController.php]: this method is duplicated

File: src/Controllers/AuthController.php (hunk 3 above)
Comment (lodry) [reply src/Controllers/AuthController.php:56]: хмм, не має працювати

File: src/MiddleWares/AuthMiddleware.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+class AuthMiddleware
+{
+	public function __construct(
+        private TokenService $tokenService,
+	) {
+
+	}
+
+    public function getCookie(string $rawCookie, string $name): string|bool
```
Comment (lodry) [inline src/MiddleWares/AuthMiddleware.php]: дублювання методу

File: server.php [hunk 6]
```diff
@@ -18,53 +24,57 @@
...
-        servePublic($publicUri, $res, $result);
-        return;
-    }
-    /**
-     * @var $router Router
-     */
-    $router = $loader->getContainer()->get('router');
-    $method = HttpMethods::tryFrom($req->method);
+    
+    $response = $req->headers["Sec-Fetch-Mode"] === 'cors'
```
Comment (lodry) [inline server.php]: better match to use

File: src/Controllers/AuthController.php [hunk 7]
```diff
@@ -1,37 +1,158 @@
...
+            }
+
+            $userData = $this->userService->register($candidate);
+            $this->logger->info('Register user', ['email' => $candidate->getEmail()]);
+            
+            // TODO resolve problem Max age
+            $maxAge = 3600 * 24 * 15;
+            $response->setHeader("Set-Cookie",'refreshToken=' . $userData->tokens->refreshToken . '; Path=/; Max-Age=' . $maxAge . '; HttpOnly');
+            $response->setHeader("Set-Cookie",'accessToken=' . $userData->tokens->accessToken . '; Path=/; Max-Age=900;');
+            $response->send(json_encode($userData));
```
Comment (lodry) [inline src/Controllers/AuthController.php]: u can encode inside response

File: server.php [hunk 8]
```diff
@@ -18,53 +24,57 @@
...
+        }
+        /**
+         * @var $router Router
+         */
+        $router = $loader->getContainer()->get('router');
+        $method = HttpMethods::tryFrom($req->method);
+
+        // TODO: Add Main error handler
+        if (!$method) {
+            throw new Exception('Incorrect method' . $req->method, 404);
```
Comment (lodry) [inline server.php:59]: leave TODO rewrite to appropriate exception classes

File: server.php [hunk 9]
```diff
@@ -18,53 +24,57 @@
...
+        }
+
+        $route = $router->matchFromPath($req->uri, $method);
+        if (!$route) {
+            throw new Exception('Path not found ' . $req->uri, 404);
```
Comment (lodry) [inline server.php:65]: leave TODO rewrite to appropriate exception classes

File: src/Controllers/AuthController.php [hunk 10]
```diff
@@ -1,37 +1,156 @@
...
+                $response->send(['title' => 'Login']);
+                return;
+            }
+
+            // var_dump($request->body);
+            $email = $request->body['email'];
+            $password = $request->body['password'];
+
+            if (!$email && !$password) {
+                throw new Exception('Incorrect data', 400);
```
Comment (lodry) [inline src/Controllers/AuthController.php]: leave TODO rewrite to appropriate exception classes

File: src/Controllers/AuthController.php [hunk 11]
```diff
@@ -1,37 +1,156 @@
...
+            }
+
+            if (!$this->userService->isAllowed($email)) {
+                throw new Exception('Access not allowed', 403);
```
Comment (lodry) [inline src/Controllers/AuthController.php]: leave TODO rewrite to appropriate exception classes

File: src/Controllers/AuthController.php [hunk 12]
```diff
@@ -1,37 +1,156 @@
...
+            }
+            
+            $user = $this->userService->findOne($email);
+            if (!$user) {
+                throw new Exception('Incorrect login, password', 404);
```
Comment (lodry) [inline src/Controllers/AuthController.php]: leave TODO rewrite to appropriate exception classes

File: src/Controllers/AuthController.php [hunk 13]
```diff
@@ -1,37 +1,156 @@
...
+            }
+            
+            if (!$this->userService->comparePassword($password, $user->getPassword())) {
+                throw new Exception('Incorrect login, password', 404);
```
Comment (lodry) [inline src/Controllers/AuthController.php]: leave TODO rewrite to appropriate exception classes

File: src/Controllers/AuthController.php [hunk 14]
```diff
@@ -1,37 +1,156 @@
...
+            $this->userService->logout($refreshToken);
+            // $this->logger->info('Logout user', ['email' => $userData->user->getEmail()]);
+            
+            // TODO rework delete cookie
+            $response->setHeader("Set-Cookie",'refreshToken=0; Path=/; Max-Age=0;');
+            $response->setHeader("Set-Cookie",'accessToken=0; Path=/; Max-Age=0;');
+            
+            $response->send(['logout'=> true]);
+        
+        } catch (\Throwable $th) {
```
Comment (lodry) [inline src/Controllers/AuthController.php]: why are you try catching twice? here and in server?

File: src/Controllers/AuthController.php [hunk 15]
```diff
@@ -1,37 +1,156 @@
...
+            $userData = $this->userService->refresh($refreshToken, '');
+            if (!$userData){
+                throw new Exception('Unauthorized', 401);
+            }
+
+            $maxAge = 3600 * 24 * 15;
+            $response->setHeader("Set-Cookie",'refreshToken=' . $userData->tokens->refreshToken . '; Path=/; Max-Age=' . $maxAge . '; HttpOnly');
+            $response->setHeader("Set-Cookie",'accessToken=' . $userData->tokens->accessToken . '; Path=/; Max-Age=900;');
+            $response->send($userData);
+        } catch (\Throwable $th) {
```
Comment (lodry) [inline src/Controllers/AuthController.php]: the same as above

File: src/Response/AbstractResponse.php [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    {
+    }
+
+    abstract protected function prepare(array|object $value);
+
+    abstract protected function error(array|object $value);
+
+    public function send(array|object|string $value){
+        if (gettype($value) === 'string') {
+            return $this->response->send($value);
```
Comment (lodry) [inline src/Response/AbstractResponse.php]: i think you should separate it in another method... like `sendRaw`

File: src/Routing/Route.php [hunk 17]
```diff
@@ -31,22 +35,55 @@ class Route
...
     }
 
-    private function __construct(string $name, string $path, callable | string  $controller, array $methods = [HttpMethods::GET], array $vars = [])
+    private function __construct(
+        string $name, 
+        string $path, 
+        callable | string  $controller, 
+        array $methods = [HttpMethods::GET], 
+        array $vars = [],
+        bool  $isRequireAuth = false
```
Comment (lodry) [inline src/Routing/Route.php]: remove space

**Review by lodry — APPROVED** [review]
almost everything is good, fix small comments

**Review by lodry — APPROVED** [review]

**Pull request merged**